src/lib.rs

```diff
@@ -1087,31 +1081,35 @@ where
 {
     fn serialize<W: borsh::io::Write>(&self, writer: &mut W) -> borsh::io::Result<()> {
         let value = self.value();
-        let length = (BITS + 7) / 8;
-        let mut bytes = 0;
-        let mask: T = u8::MAX.into();
-        while bytes < length {
-            let le_byte: u8 = ((value >> (bytes << 3)) & mask)
+        let total_bytes = (BITS + 7) / 8;
+        let mut byte_count_written = 0;
+        let byte_mask: T = u8::MAX.into();
+        while byte_count_written < total_bytes {
+            let le_byte: u8 = ((value >> (byte_count_written << 3)) & byte_mask)
                 .try_into()
                 .ok()
                 .expect("we cut to u8 via mask");
             writer.write(&[le_byte])?;
-            bytes += 1;
+            byte_count_written += 1;
         }
         Ok(())
     }
 }
 
 #[cfg(feature = "borsh")]
 impl<
-        T: BorshDeserialize + core::cmp::PartialOrd<<UInt<T, BITS> as Number>::UnderlyingType>,
+        T: borsh::BorshDeserialize + PartialOrd<<UInt<T, BITS> as Number>::UnderlyingType>,
         const BITS: usize,
-    > BorshDeserialize for UInt<T, BITS>
+    > borsh::BorshDeserialize for UInt<T, BITS>
 where
     Self: Number,
 {
     fn deserialize_reader<R: borsh::io::Read>(reader: &mut R) -> borsh::io::Result<Self> {
-        let mut buf = vec![0u8; core::mem::size_of::<T>()];
+        // Ideally, we'd want a buffer of size `BITS >> 3` or `size_of::<T>`, but that's not possible
+        // with arrays.
+        // So instead we'll do a 16 byte buffer which handles the largest arbitrary-ints possible -
+        // not ideal, but still pretty small and better than going through an allocator.
+        let mut buf = [0u8; 16];
```

Comment:
hmm this might be wrong - we might be asking too much from Borsh here. Will add a test to catch this and fix the implementation